### DISCUSSION TITLE
Possible Windows release of disk-utilities

### DISCUSSION THREAD

**Discussion by realmml** [keirf/disk-utilities#133] [Ideas]
Hi! :)

Although I use Linux, for a few friends is more easy to being able to use Windows.

For that case, I have compiled a windows version for them, because they are not so technical mind like me. Making possible that everybody can help with the preservation task. 

I know that there is an issue open where somebody is asking for a pre-build windows binary and in the eab thread acd2001 usually upload windows executables. Then I only would like to give my vote for at least a windows version for when you make a release.

Thanks!
Mauri.

**Comment by keirf** [comment]
Great idea. If you know how to build one, it could be integrated into GitHub workflows, which is how I already automate builds of Greaseweazle (for example). Perhaps you would fancy having a go at that? It's a special YAML file living under .github/workflows. Or if you give me some command line runes and prerequisites for building on Windows, perhaps I/we could work out the YAML.

**Reply by realmml** [reply]
Thanks, I will try to take a look during the weekend :)

**Comment by realmml** [comment]
Hi! :)

Today, I got some time for taking a look to my patches for being able to compile the disk-analyse in windows using msys2/mingw64. 

Let me explain the patches and I will try to make a diff and look the github/workflows when I can.

The only source file that we need to change is libdisk/stream)/caps.c, and the main problem is that windows has not dlfcn.h. Of course, you can find a few github repositories with dlfcn implementations  for windows; but in my case, being a Linux user, I preferred to use a more portable way, then I used the [GNU Libtool] (https://www.gnu.org/software/libtool/).

And only changed a few lines in the original:
```
/*
 * stream/caps.c
 */
.
.
.
#include <dlfcn.h>

#ifdef __APPLE__
#define CAPSLIB_NAME    "/Library/Frameworks/CAPSImage.framework/CAPSImage"
#else
#define CAPSLIB_NAME    "libcapsimage.so"
#endif
.
.
.
.
.
.
static int get_capslib(void)
{
    if (capslib.ref++)
        return 1;

#ifdef __APPLE__
    if ((capslib.handle = dlopen(CAPSLIB_NAME, RTLD_LAZY)) == NULL) {
        warnx("Unable to open " CAPSLIB_NAME);
        goto fail_no_handle;
    }
    capslib.version = 5; /* guess */
#else
    if ((capslib.handle = dlopen(CAPSLIB_NAME ".5", RTLD_LAZY))) {
        capslib.version = 5;
    } else if ((capslib.handle = dlopen(CAPSLIB_NAME ".4", RTLD_LAZY))) {
        capslib.version = 4;
    } else {
        warnx("Unable to open " CAPSLIB_NAME ".5 or " CAPSLIB_NAME ".4");
        goto fail_no_handle;
    }
#endif

#define GETSYM(sym)                                     \
    capslib.sym = dlsym(capslib.handle, "CAPS"#sym);    \
   if (dlerror() != 0) goto fail;
.
.
.
.
.
.
fail:
    dlclose(capslib.handle);
fail_no_handle:
        print_library_download_info();
        --capslib.ref;
    } // lt_dlinit() failed!
    return 0;
}

static void put_capslib(void)
{
    if (--capslib.ref)
        return;
    CAPSExit();
    dlclose(capslib.handle);
}
.
.
.
```
To:
```/*
 * stream/caps.c
 */
.
.
.
#include <ltdl.h> // instead of #include <dlfcn.h>

#ifdef __APPLE__
#define CAPSLIB_NAME    "/Library/Frameworks/CAPSImage.framework/CAPSImage"
#elif __linux__
#define CAPSLIB_NAME    "libcapsimage.so"
#else // Windows
#define CAPSLIB_NAME    "CAPSImg.dll"
#endif
.
.
.
.
.
.
static int get_capslib(void)
{   
    if (lt_dlinit()) { // Initialize the libdltl

        if (capslib.ref++)
            return 1;

#ifdef __APPLE__
        if ((capslib.handle = dlopen(CAPSLIB_NAME, RTLD_LAZY)) == NULL) {
            warnx("Unable to open " CAPSLIB_NAME);
            goto fail_no_handle;
        }
        capslib.version = 5; /* guess */
#elif __linux__
        if ((capslib.handle = dlopen(CAPSLIB_NAME ".5", RTLD_LAZY))) {
            capslib.version = 5;
        } else if ((capslib.handle = dlopen(CAPSLIB_NAME ".4", RTLD_LAZY))) {
            capslib.version = 4;
        } else {
            warnx("Unable to open " CAPSLIB_NAME ".5 or " CAPSLIB_NAME ".4");
            goto fail_no_handle;
        }
#else // Windows
        if ((capslib.handle = lt_dlopen(CAPSLIB_NAME)) == NULL) { // Uses RTLD_LAZY by default
            warnx("Unable to open " CAPSLIB_NAME);
            goto fail_no_handle;
        }
        capslib.version = 5; //CAPS_LIB_RELEASE;
#endif

#define GETSYM(sym)                                     \
        capslib.sym = lt_dlsym(capslib.handle, "CAPS"#sym);    \
        if (lt_dlerror() != 0) goto fail;
.
.
.
fail:
        lt_dlclose(capslib.handle);
    
fail_no_handle:
        print_library_download_info();
        --capslib.ref;
    } // lt_dlinit() failed!
    return 0;
}

static void put_capslib(void)
{
    if (--capslib.ref)
        return;
    CAPSExit();
    lt_dlclose(capslib.handle);
}
.
.
.
```
As you can see, I only had to include the new header file (ltdl.h), add the extra call lt_dlinit for initialize the library and rename the dlxxx functions by lt_dlxxx and that it was enough.

Of course, being a windows program I compiled everything static for not needing extra DLLs, that means that I changed the library parameters in the makefile for libdisk from:
```LIBS-$(caps) := -ldl```
To:
```LIBS-$(caps) := -lltdl -static-libgcc -static-libstdc++ -static ```

And in the disk-analyse makefile from:
```LIBS := -L../libdisk -ldisk```
To:
```LIBS := -L../libdisk -ldisk -lltdl -static-libgcc -static-libstdc++ -static```

Aside of using static , I need to link with the ltdl library instead of dl.

And that is all, of course I should static compile the rest of disk-utilities, ... but as you can see the most traumatic change was to change from dl to gnu libtool.
